### PR TITLE
`Adaptive learning`: Simplify date picker for competency due dates to exclude hours and minutes

### LINKED ISSUE #9678: Reduce date picker for competency due dates to select only days
### Is your feature request related to a problem?

The configuration of competencies should be simpler.

### Describe the solution you'd like

Remove the option to select hours and minutes in the date picker when setting the due date of a competency.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Resolves #9678 

### Description

Improves our date picker by allowing developers to chose a picker that allows users to select
- only time
- only date
- both (still the default)

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a competency and select a due date
4. The date picker should only let you select the day, not time

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

unchanged

### Screenshots

Screenshot 2024-11-08 at 20 45 35

## Summary by CodeRabbit

- **New Features**
	- Enhanced date-time picker functionality with the addition of a calendar option.
	- Improved user interface with dynamic button rendering based on selected date-time picker type.

- **Bug Fixes**
	- Streamlined error handling and validation feedback for user inputs in the date-time picker.

- **Documentation**
	- Updated component structure for clarity and improved functionality of date-time picker instances.

### CODE REVIEW

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Works as described

image

**Review by sawys777 — DISMISSED** [review]
testingCompe
Tested on TS1, everything worked accordingly to the test description.

**Review by flbrgit — APPROVED** [review]
Tested on TS2, everything works as described, only dates are pickable. One question, though: Does it make much sense to be able to pick dates in the past when creating a new competency?
Screenshot 2024-11-09 111808

File: src/main/webapp/app/shared/date-time-picker/date-time-picker.component.ts [hunk 1]
```diff
@@ -118,4 +125,6 @@ export class FormDateTimePickerComponent implements ControlValueAccessor {
     clearDate() {
         this.dateInput.reset(undefined);
     }
+
+    protected readonly DateTimePickerType = DateTimePickerType;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/shared/date-time-picker/date-time-picker.component.ts:129]: We usually have this at the top of the file

File: src/main/webapp/app/shared/date-time-picker/date-time-picker.component.ts [hunk 2]
```diff
@@ -28,6 +34,7 @@ export class FormDateTimePickerComponent implements ControlValueAccessor {
     @Input() min?: dayjs.Dayjs; // Dates before this date are not selectable.
     @Input() max?: dayjs.Dayjs; // Dates after this date are not selectable.
     @Input() shouldDisplayTimeZoneWarning = true; // Displays a warning that the current time zone might differ from the participants'.
+    @Input() pickerType = DateTimePickerType.DEFAULT; // Only show the date picker, not the time picker.
```
Comment (JohannesStoehr) [inline src/main/webapp/app/shared/date-time-picker/date-time-picker.component.ts:37]:
```suggestion
    @Input() pickerType = DateTimePickerType.DEFAULT; // Show the date and time picker
```
I think this comment is wrong, isn't it?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by sachmii — APPROVED** [review]
Tested on TS2, works as described.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by sawys777 — DISMISSED** [review]
Tested on TS1, everything worked as described

**Review by sachmii — DISMISSED** [review]
Tested on TS1, re-approve.

**Review by HawKhiem — APPROVED** [review]
Tested on TS1. Reapprove

**Review by JohannesWt — APPROVED** [review]
Code

**Review by Haoyuli2002 — COMMENTED** [review]
Tested on TS3 as user 17(instructor), everything works well, we could only select the date, not the time anymore.

File: src/main/webapp/app/shared/date-time-picker/date-time-picker.component.html [hunk 3]
```diff
@@ -26,19 +26,35 @@
...
+            <button [owlDateTimeTrigger]="dtCalendar" class="btn position-absolute" type="button">
+                <fa-icon [icon]="faCalendarAlt" />
+            </button>
+        }
+        @default {
+            <button [owlDateTimeTrigger]="dtDefault" class="btn position-absolute" type="button">
+                <fa-icon [icon]="faCalendarAlt" />
+            </button>
+        }
+    }
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/date-time-picker/date-time-picker.component.html:48]: Optional: 

 ```
@switch (pickerType()) {
        @case (DateTimePickerType.TIMER) {
            <ng-template *ngTemplateOutlet="owlTrigger; context: { $implicit: dtTimer, icon: faClock}" />
        }
        @case (DateTimePickerType.CALENDAR) {
            <ng-template *ngTemplateOutlet="owlTrigger; context: { $implicit: dtCalendar, icon: faCalendarAlt}" />
        }
        @default {
           <ng-template *ngTemplateOutlet="owlTrigger; context: { $implicit: dtDefault, icon: faCalendarAlt}" />
        }
    }
    
    <ng-template #owlTrigger let-owlTrigger let-icon="icon"> 
     <button [owlDateTimeTrigger]="owlTrigger" class="btn position-absolute" type="button">
                <fa-icon [icon]="icon" />
     </button>
    </ng-template> 
```

Has the advantage that if you change something on the button (classes, etc.), you only have to adjust it in one place.
Disclaimer: Possible typos. Please double-check when copying and pasting.

**Review by zagemello — DISMISSED** [review]
image
Tested on TS3, works as expected

**Review by kevinfischer4 — DISMISSED** [review]
Tested on TS3, works as described.
Screenshot 2024-11-11 at 13 28 05

**Review by raffifasaro — DISMISSED** [review]
Code

**Comment by MaximilianAnzinger** [comment]
@rabeatwork i honestly don't know what kind of dark magic this was, but I like it 👍 
Picasso! 🎨

**Review by JohannesWt — APPROVED** [review]
Code reapprove

**Review by asliayk — APPROVED** [review]
tested locally and works as expected

**Review by vinceclifford — APPROVED** [review]
Tested on TS5, works as expected

**Review by sachmii — APPROVED** [review]
Tested on TS5, re-approve.

**Review by rabeatwork — APPROVED** [review]

**Pull request merged**